### DISCUSSION TITLE
Using returned values form a trial function in another trial function

### DISCUSSION THREAD

**Discussion by Alexanderfrs** [jspsych/jsPsych#1105] [Q&A] [answered]
Hi everyone,

I'm at my first attempt of creating a simple experiment with JsPsych and got stuck at trying to access returned values in different functions. Specifically, my goal is to let participants decide between two gambling options that have randomly generated possible outcomes (and probabilities of these outcomes occurring) at each trial. 

For this, I set up a function named 'generateRandomStimulus' that generates and returns these parameters for a given trial, and I assigned the function to be the 'stimulus:' part of a timeline trial. After participants make a choice between the two gambles, the next timeline trial I would want to include is a feedback trial that grabs the possible outcomes generated in the previous trial's function, informs the participant of the outcome of the gamble that they have chosen, and add the outcome to a total score that is kept track of across the entire experiment. 

Everything I tried this far (like calling the first function and assign its returned values to global variables to use them in the second function/trial) didn't work. Thus, any hints or efficient ways to go about this will be highly appreciated! Here is my code:

`       

        // create an empty timeline array
        var timeline = [];

        // current score 
        var currentScore = 0

        // create trial
        var trial = {
            type: 'html-button-response',
            stimulus: generateRandomStimulus(),
            choices: ['Gamble 1', 'Gamble 2'],
            prompt: "<p>How do you choose?</p>",
            post_trial_gap: 200
        }
        timeline.push(trial);

        var feedback_trial = {
            type: 'html-keyboard-response',
            stimulus: feedbackFunc
        }
        timeline.push(feedback_trial)

        // run the experiment
        jsPsych.init({
            timeline: timeline
        });

        // gamble function
        function generateRandomStimulus() { 
            // Step 1. generate all possible outcome values and their respective probabilities for a trial 
            var max_outcome = 10;
            var min_outcome = -10;
            var x1 = Math.floor(Math.random() * (max_outcome - min_outcome + 1) + min_outcome);
            var y1 = Math.floor(Math.random() * (max_outcome - min_outcome + 1) + min_outcome);
            var x2 = Math.floor(Math.random() * (max_outcome - min_outcome + 1) + min_outcome);
            var y2 = Math.floor(Math.random() * (max_outcome - min_outcome + 1) + min_outcome);
            var max_proba = 1;
            var min_proba = 0.5;
            var p1 = Math.random() * (1 - min_proba) + min_proba;
            p1 = Math.round(p1 * 100) / 100;
            var p2 = 1 - p1;
            p2 = Math.round(p2 * 100) / 100;

            // Step 2. include the outcomes of each gamble and the probabilities in separate arrays
            var arr_gamble1 = [x1, y1];
            var arr_gamble2 = [x2, y2];
            var arr_proba = [p1, p2];

            // Step 3. Present them and let participant make a choice
            return "GAMBLE 1: " + p1 * 100 + " percent chance to win " + x1 + " or " + p2 * 100 + " percent chance to win " + y1 + " points. " +
                    "GAMBLE 2: " + p1 * 100 + " percent chance to win " + x2 + " or " + p2 * 100 + " percent chance to win " + y2 + " points."
        }
         // feedback function and score update   
        function feedbackFunc() {
            var last_button_pressed = jsPsych.data.getLastTrialData().values()[0].button_pressed;
            if (last_button_pressed == 0) {
                var outcome_gamble1 = jsPsych.randomization.sampleWithReplacement(arr_gamble1, 1, arr_proba)
                current_score += outcome_gamble1
                return "The outcome of gamble 1 was " + outcome_gamble1 + " your updated score is " + current_score
            } else {
                var outcome_gamble2 = jsPsych.randomization.sampleWithReplacement(arr_gamble2, 1, arr_proba)
                current_score += outcome_gamble2
                return "The outcome of gamble 2 was " + outcome_gamble2 + " your updated score is " + current_score
            }
        }
            
        
    `

Cheers

**Comment by becky-gilbert** [comment, marked as answer]
Hi @Alexanderfrs, this general approach makes sense to me. 

One reason your code isn't working is that there are three variables that you define inside of a function (`generateRandomStimulus`) but then try to use inside of a different function (`feedbackFunc`). These variables are `arr_gamble1`, `arr_gamble2` and `arr_proba`. This produces an error in `feedbackFunc`: uncaught reference error <variable> not defined. Any variables that are defined within a function (i.e. set up using a keyword like `var`, `let` or `const`) are only available inside that function. So the trick is to define them in your global scope, as you do with the current score value. You can either leave the value undefined like this:
```javascript
var arr_gamble1, arr_gamble2, arr_proba;
```
Or initialize them with a value, if that makes more sense for what you're doing.

A second issue is that the function [`jsPsych.randomization.sampleWithReplacement`](https://www.jspsych.org/core_library/jspsych-randomization/#jspsychrandomizationsamplewithreplacement) returns an array. So to get the value out of that array, you need to index the array with 0 (i.e. the first/only value in the array, since you're just sampling one element).
```javascript
var outcome_gamble1 = jsPsych.randomization.sampleWithReplacement(arr_gamble1, 1, arr_proba)[0]
```
A third problem is that you've initialized the variable to hold the current score as `currentScore`, but then try to set and access its value using `current_score`.

**Reply by Alexanderfrs** [reply]
Hi @becky-gilbert, an infinite thanks for taking the time to help! It absolutely makes sense, it runs without problems now. :)
